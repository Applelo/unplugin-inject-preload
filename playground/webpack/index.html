<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title><%= htmlWebpackPlugin.options.title %></title>
    <!--__unplugin-inject-preload__-->
    <%= htmlWebpackPlugin.tags.headTags %>
  </head>
  <body>
    <h1><%= htmlWebpackPlugin.options.title %></h1>
    <script type="module" src="./main.ts"></script>
  </body>
</html>