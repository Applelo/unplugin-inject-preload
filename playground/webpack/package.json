{
  "private": true,
  "keywords": [],
  "scripts": {
    "build": "webpack --config webpack.config.js --mode production"
  },
  "devDependencies": {
    "css-loader": "^6.8.1",
    "html-webpack-plugin": "^5.5.3",
    "mini-css-extract-plugin": "^2.7.6",
    "ts-loader": "^9.5.1",
    "typescript": "^5.2.2",
    "unplugin-inject-preload": "workspace:*",
    "webpack": "^5.89.0",
    "webpack-cli": "^5.1.4"
  }
}
