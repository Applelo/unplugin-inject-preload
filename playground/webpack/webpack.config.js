const path = require('node:path')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const UnpluginInjectPreload = require('unplugin-inject-preload/webpack').default

module.exports = {
  entry: path.resolve(__dirname, './../src/main.ts'),
  output: {
    publicPath: 'dist',
    filename: 'main.js',
    path: path.resolve(__dirname, 'dist'),
    clean: true,
    assetModuleFilename: '[name].[hash][ext][query]',
  },
  resolve: {
    extensions: ['.ts', '.js'],
    extensionAlias: {
      '.js': ['.js', '.ts'],
    },
  },
  module: {
    rules: [
      {
        test: /\.css$/i,
        use: [MiniCssExtractPlugin.loader, 'css-loader'],
      },
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/i,
        type: 'asset/resource',
      },
      // all files with a `.ts`, `.cts`, `.mts` or `.tsx` extension will be handled by `ts-loader`
      {
        test: /\.([cm]?ts|tsx)$/,
        loader: 'ts-loader',
      },
    ],
  },
  plugins: [
    new MiniCssExtractPlugin(),
    new HtmlWebpackPlugin({
      inject: false,
      minify: false,
      templateContent: ({ htmlWebpackPlugin }) => `
      <!DOCTYPE html>
      <html>
        <head>
          <!--__unplugin-inject-preload__-->
          ${htmlWebpackPlugin.tags.headTags}
        </head>
        <body>
          <h1>Hello World</h1>
          ${htmlWebpackPlugin.tags.bodyTags}
        </body>
      </html>
    `,
    }),
    UnpluginInjectPreload({
      injectTo: 'custom',
      files: [
        {
          outputmatch: /Roboto-[a-zA-Z]*.[a-z-0-9]*\.woff2$/,
        },
        {
          outputMatch: /^(?!main).*\.(css|js)$/,
        },
      ],
    }),
  ],
}
