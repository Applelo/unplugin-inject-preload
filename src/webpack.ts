/**
 * This entry file is for Webpack plugin.
 *
 * @module
 */

import unplugin from './index'

/**
 * Webpack plugin
 *
 * @example
 *```ts
 * // webpack.config.js
 * const HtmlWebpackPlugin = require('html-webpack-plugin')
 * const UnpluginInjectPreload = require('unplugin-inject-preload/webpack')
 *
 * module.exports = {
 *   plugins: [
 *     HtmlWebpackPlugin({ your_options }),
 *     UnpluginInjectPreload({ your_options }),
 *   ]
 * }
 *```
 */
export default unplugin.webpack
