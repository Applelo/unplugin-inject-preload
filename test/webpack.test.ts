import { join } from 'node:path'
import { readFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'
import type { Configuration } from 'webpack'
import webpack from 'webpack'
import HtmlWebpackPlugin from 'html-webpack-plugin'
import MiniCssExtractPlugin from 'mini-css-extract-plugin'
import UnpluginInjectPreload from './../src/webpack'
import type { Options } from './../src/types'
import configs from './fixtures/configs'

async function buildWebpack(pluginConfig: Options, config: Configuration = {}) {
  return new Promise(((resolve, reject) => {
    const compiler = webpack({
      mode: 'production',
      context: join(__dirname, 'fixtures/webpack'),
      entry: join(__dirname, 'fixtures/src/main.ts'),
      output: {
        path: join(__dirname, 'fixtures/webpack/dist'),
        publicPath: config.output?.publicPath || '/',
        clean: true,
        assetModuleFilename: '[name]-[hash][ext][query]',
      },
      resolve: {
        extensions: ['.ts', '.js'],
        extensionAlias: {
          '.js': ['.js', '.ts'],
        },
      },
      module: {
        rules: [
          {
            test: /\.css$/i,
            use: [MiniCssExtractPlugin.loader, 'css-loader'],
          },
          {
            test: /\.(woff|woff2|eot|ttf|otf)$/i,
            type: 'asset/resource',
          },
          // all files with a `.ts`, `.cts`, `.mts` or `.tsx` extension will be handled by `ts-loader`
          {
            test: /\.([cm]?ts|tsx)$/,
            loader: 'ts-loader',
          },
        ],
      },
      plugins: [
        new MiniCssExtractPlugin(),
        new HtmlWebpackPlugin({
          minify: false,
          inject: false,
          template: join(__dirname, 'fixtures/webpack/index.html'),
        }),
        UnpluginInjectPreload(pluginConfig),
      ],
    })

    compiler.run((err, stats) => {
      if (err)
        reject(err)

      const statsErrors = stats ? stats.compilation.errors : []
      if (statsErrors.length > 0) {
        console.error(statsErrors)
        reject(statsErrors)
      }

      const result = readFileSync(
        join(__dirname, 'fixtures/webpack/dist/index.html'),
        'utf8',
      )

      resolve(result)
    })
  }))
}

describe('excerpt webpack', () => {
  for (const key in configs) {
    if (Object.prototype.hasOwnProperty.call(configs, key)) {
      const config = configs[key] as Options
      it(`test ${key}`, async () => {
        const output = await buildWebpack(config)
        expect(output).toMatchSnapshot()
      })

      it(`test ${key} with basePath`, async () => {
        const output = await buildWebpack(config, { output: { publicPath: '/base' } })
        expect(output).toMatchSnapshot()
      })
    }
  }
})
