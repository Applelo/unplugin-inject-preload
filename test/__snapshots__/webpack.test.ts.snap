// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`excerpt webpack > test auto 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test auto with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/base/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/base/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test customAttributes 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\" crossorigin=\\"anonymous\\" data-font=\\"Roboto\\"><link rel=\\"preload\\" href=\\"/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\" crossorigin=\\"anonymous\\" data-font=\\"Roboto\\"><script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test customAttributes with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/base/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\" crossorigin=\\"anonymous\\" data-font=\\"Roboto\\"><link rel=\\"preload\\" href=\\"/base/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\" crossorigin=\\"anonymous\\" data-font=\\"Roboto\\"><script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test customPosition 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">

    <link rel=\\"preload\\" href=\\"/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\" data-vite-plugin-inject-preload>
    <link rel=\\"preload\\" href=\\"/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\" data-vite-plugin-inject-preload>

  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test customPosition with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">

    <link rel=\\"preload\\" href=\\"/base/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\" data-vite-plugin-inject-preload>
    <link rel=\\"preload\\" href=\\"/base/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\" data-vite-plugin-inject-preload>

  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test injectBottom 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\"><link rel=\\"preload\\" href=\\"/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test injectBottom with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\"><link rel=\\"preload\\" href=\\"/base/Roboto-Italic-96e486355f319e0b182b.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/base/Roboto-Regular-ec5068bd8823a38af494.woff2\\" type=\\"font/woff2\\" as=\\"font\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test modulepreload 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test modulepreload with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test noType 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/Roboto-Regular-ec5068bd8823a38af494.woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/Roboto-Italic-96e486355f319e0b182b.woff2\\" as=\\"font\\"><script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test noType with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"/base/Roboto-Regular-ec5068bd8823a38af494.woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"/base/Roboto-Italic-96e486355f319e0b182b.woff2\\" as=\\"font\\"><script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test wrongAttributes 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"./yolo.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"./yolo.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><script defer src=\\"/main.js\\"></script><link href=\\"/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;

exports[`excerpt webpack > test wrongAttributes with basePath 1`] = `
"<!DOCTYPE html>
<html>
  <head>
    <meta charset=\\"utf-8\\"/>
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\" />
    <title>Webpack App</title>
    <link rel=\\"preload\\" href=\\"./yolo.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><link rel=\\"preload\\" href=\\"./yolo.woff2\\" type=\\"font/woff2\\" as=\\"font\\"><script defer src=\\"/base/main.js\\"></script><link href=\\"/base/main.css\\" rel=\\"stylesheet\\">
    <!--__unplugin-inject-preload__-->
  </head>
  <body>
    <h1>Unplugin Inject Preload</h1>
    
  </body>
</html>
"
`;
